File: pt5/1152.analyze-user-website-visit-pattern.go
package main

import "sort"

type visitRecord struct {
	user string
	time int
	site string
}

func mostVisitedPattern(username []string, timestamp []int, website []string) []string {
	visits := make([]visitRecord, len(username))
	for i := range username {
		visits[i] = visitRecord{user: username[i], time: timestamp[i], site: website[i]}
	}
	sort.Slice(visits, func(i, j int) bool {
		return visits[i].time < visits[j].time
	})

	byUser := map[string][]string{}
	for _, visit := range visits {
		byUser[visit.user] = append(byUser[visit.user], visit.site)
	}

	counts := map[[3]string]int{}
	for _, sites := range byUser {
		seen := map[[3]string]bool{}
		for i := 0; i < len(sites); i++ {
			for j := i + 1; j < len(sites); j++ {
				for k := j + 1; k < len(sites); k++ {
					pattern := [3]string{sites[i], sites[j], sites[k]}
					seen[pattern] = true
				}
			}
		}
		for pattern := range seen {
			counts[pattern]++
		}
	}

	if len(counts) == 0 {
		return []string{}
	}

	var best [3]string
	bestCount := -1
	first := true
	for pattern, count := range counts {
		if first || count > bestCount || count == bestCount && lessPattern(pattern, best) {
			first = false
			best = pattern
			bestCount = count
		}
	}

	return []string{best[0], best[1], best[2]}
}

func lessPattern(a, b [3]string) bool {
	for i := 0; i < 3; i++ {
		if a[i] != b[i] {
			return a[i] < b[i]
		}
	}
	return false
}

File: pt5/1152.analyze-user-website-visit-pattern.java
import java.util.*;

class Solution {
    public List<String> mostVisitedPattern(String[] username, int[] timestamp, String[] website) {
        Integer[] order = new Integer[username.length];
        for (int i = 0; i < order.length; i++) {
            order[i] = i;
        }
        Arrays.sort(order, Comparator.comparingInt(i -> timestamp[i]));

        Map<String, List<String>> visits = new HashMap<>();
        for (int index : order) {
            visits.computeIfAbsent(username[index], key -> new ArrayList<>()).add(website[index]);
        }

        Map<List<String>, Integer> patternCount = new HashMap<>();
        for (List<String> sites : visits.values()) {
            Set<List<String>> seen = new HashSet<>();
            for (int i = 0; i < sites.size(); i++) {
                for (int j = i + 1; j < sites.size(); j++) {
                    for (int k = j + 1; k < sites.size(); k++) {
                        seen.add(Arrays.asList(sites.get(i), sites.get(j), sites.get(k)));
                    }
                }
            }
            for (List<String> pattern : seen) {
                patternCount.put(pattern, patternCount.getOrDefault(pattern, 0) + 1);
            }
        }

        List<String> best = new ArrayList<>();
        int bestCount = -1;
        for (Map.Entry<List<String>, Integer> entry : patternCount.entrySet()) {
            List<String> pattern = entry.getKey();
            int count = entry.getValue();
            if (count > bestCount || count == bestCount && lexicographicallySmaller(pattern, best)) {
                best = pattern;
                bestCount = count;
            }
        }

        return best;
    }

    private boolean lexicographicallySmaller(List<String> a, List<String> b) {
        if (b.isEmpty()) {
            return true;
        }
        for (int i = 0; i < 3; i++) {
            int cmp = a.get(i).compareTo(b.get(i));
            if (cmp != 0) {
                return cmp < 0;
            }
        }
        return false;
    }
}

